Update JWT guide with the builder API intro and fix some typos (#7351)
quarkusio/quarkus: docs/src/main/asciidoc/security-jwt.adoc

=== Generating a JWT

> Often one obtains a JWT from an identity manager like https://www.keycloak.org/[Keycloak], but for this quickstart we will generate our own using the JWT generation API provided by `smallrye-jwt` (see <<Create JWT tokens with Smallrye JWT>> for more infomation) and the TokenUtils class shown in the following listing. Take this source and place it into `security-jwt-quickstart/src/test/java/org/acme/jwt/TokenUtils.java`.

Reviewer: Please use an id to make the link. If you change the title, the link will be broken.

|smallrye.jwt.token.header|`Authorization`|Set this property if another header such as `Cookie` is used to pass the token.
|smallrye.jwt.token.cookie|none|Name of the cookie containing a token. This property will be effective only if  `smallrye.jwt.token.header` is set to `Cookie`.
|smallrye.jwt.token.schemes|`Bearer`|Comma-separated list containing an alternative single or multiple schemes, for example, `DPoP`.
|smallrye.jwt.token.kid|none|Key identifier. If it is set then the verification JWK key as well every JWT token must have a matching 'kid' header.

> |smallrye.jwt.time-to-live|none|The maximum number of seconds that a JWT may be issued for use. Effectively, the difference  between the expiration date of the JWT and the issued at date must not exceed this value.

Reviewer 2: ```suggestion
|smallrye.jwt.time-to-live|none|The maximum number of seconds that a JWT may be issued for use. Effectively, the difference between the expiration date of the JWT and the issued at date must not exceed this value.
```

|smallrye.jwt.groups-separator|' '|Separator for splitting a string which may contain multiple group values. It will only be used if the `smallrye.jwt.path.groups` property points to a custom claim whose value is a string. The default value is a single space because a standard OAuth2 `scope` claim may contain a space separated sequence.
|smallrye.jwt.claims.groups|none| This property can be used to set a default groups claim value when the current token has no standard or custom groups claim available.
|smallrye.jwt.jwks.refresh-interval|60|JWK cache refresh interval in minutes. It will be ignored unless the `mp.jwt.verify.publickey.location` points to the HTTPS URL based JWK set and no HTTP `Cache-Control` response header with a positive `max-age` parameter value is returned from a JWK HTTPS endpoint.
|smallrye.jwt.expiration.grace|60|Expiration grace in seconds. By default an expired token will still be accepted if the current time is no more than 1 min after the token expiry time.
|smallrye.jwt.verify.aud|none|Comma separated list of the audiences that a token `aud` claim may contain.
|===

== Generate JWT tokens with Smallrye JWT

> JWT claims can be signed or encrypted or signed first and the nested JWT token - encrypted.

Reviewer: ```suggestion
JWT claims can be signed or encrypted or signed first and the nested JWT token encrypted.
```

> Signing the claims is used most ofthen to secure the claims. What is known today as a JWT token is typically produced by signing the claims in a JSON format using the steps described in the `JSON Web Signature RFC7515` specification.

Reviewer 2: ```suggestion
Signing the claims is used most often to secure the claims. What is known today as a JWT token is typically produced by signing the claims in a JSON format using the steps described in the `JSON Web Signature RFC7515` specification.
```

> However, when the claims are sensitive, their confidenatiality can be guaranteed by following the steps described in the `JSON Web Encryption RFC7516` specification to produce a JWT token with the encrypted claims.

Reviewer 2: ```suggestion
However, when the claims are sensitive, their confidentiality can be guaranteed by following the steps described in the `JSON Web Encryption RFC7516` specification to produce a JWT token with the encrypted claims.
```

Reviewer 2: Also a link to the `JSON Web Encryption RFC7516` spec would be nice

Finally both the confidentiality and integrity of the claims can be further enforced by signing them first and then encrypting the nested JWT token.

> Smallrye JWT provides an API for securing the JWT claims using all of these options.

Reviewer: ```suggestion
SmallRye JWT provides an API for securing the JWT claims using all of these options.
```

> === Create JwtClaimsBuilder and set the claims.

Reviewer: ```suggestion
=== Create JwtClaimsBuilder and set the claims
```

> The first step is to initialize `JwtClaimsBuilder` using one of the options below and add some claims to it:

Reviewer: ```suggestion
The first step is to initialize a `JwtClaimsBuilder` using one of the options below and add some claims to it:
```

[source, java]
----
import java.util.Collections;
import io.smallrye.jwt.build.Jwt;
import io.smallrye.jwt.build.JwtClaimsBuilder;
...
// Create an empty builder and add some claims
JwtClaimsBuilder builder1 = Jwt.claims();
builder1.claim("customClaim", "custom-value").issuer("https://issuer.org");

// Builder created from the existing claims
JwtClaimsBuilder builder2 = Jwt.claims("/tokenClaims.json");

> // Builder created from the map of claims

Reviewer: ```suggestion
// Builder created from a map of claims
```

JwtClaimsBuilder builder3 = Jwt.claims(Collections.singletonMap("customClaim", "custom-value"));
----

> The API is fluent so the builder initialization can be done as part of the fluent API sequence. The builder will also set an `iat` (issued at) to the current time, `exp` (expires at) to 5 mins away from the current time and `jti` (unique token identifier) claims if they have not already been set, so one can skip setting them when possible.

Reviewer: ```suggestion
The API is fluent so the builder initialization can be done as part of the fluent API sequence. The builder will also set `iat` (issued at) to the current time, `exp` (expires at) to 5 minutes away from the current time and `jti` (unique token identifier) claims if they have not already been set, so one can skip setting them when possible.
```

// Set the headers and sign the claims with an RSA private key loaded in the code (the implementation of this method is omitted). Note a 'jws()' transition to a 'JwtSignatureBuilder'.
String jwt2 = Jwt.claims("/tokenClaims.json").jws().signatureKeyId("kid1").header("custom-header", "custom-value").sign(getPrivateKey());
----

Note the `alg` (algorithm) header is set to `RS256` by default.

=== Encrypt the claims

The claims can be encrypted immediately or after the `JSON Web Encryption` headers have been set the same way as they can be signed.
The only minor difference is that encrypting the claims always requires a `jwe()` `JwtEncryptionBuilder` transition:

[source, java]
----
import io.smallrye.jwt.build.Jwt;
...

// Encrypt the claims using the public key loaded from the location set with a 'smallrye.jwt.encrypt.key-location' property.
String jwt1 = Jwt.claims("/tokenClaims.json").jwe().encrypt();

// Set the headers and encrypt the claims with an RSA public key loaded in the code (the implementation of this method is omitted).
String jwt2 = Jwt.claims("/tokenClaims.json").jwe().header("custom-header", "custom-value").encrypt(getPublicKey());
----

> Note the `alg` (key management algorithm) header is set to `RSA-OAEP-256` (it will be changed to `RSA-OAEP` in the future version of smallrye-jwt) and the `enc` (content encryption header) is set to `AGCM256` by default.

Reviewer: ```suggestion
Note the `alg` (key management algorithm) header is set to `RSA-OAEP-256` (it will be changed to `RSA-OAEP` in a future version of smallrye-jwt) and the `enc` (content encryption header) is set to `AGCM256` by default.
```
